gh-97696 Add documentation for get_coro() behavior with eager tasks

jbower-fb opened the pull request:
This PR caught the plague and had to be killed with fire. A new one was opened to replace it https://github.com/python/cpython/pull/104304.

In response to comments on this issue.

<!-- gh-issue-number: gh-97696 -->
* Issue: gh-97696
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--104189.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

jbower-fb committed:
gh-97696 Add documentation for get_coro() behavior with eager tasks

gaogaotiantian commented on Doc/library/asyncio-task.rst:
```diff
@@ -1153,6 +1153,11 @@ Task Object
       Return the coroutine object wrapped by the :class:`Task`.
 
+      .. note::
+        
```
A trailing space here :)

sunmy2019 committed:
gh-104190: fix ubsan crash (#104191)

iritkatriel committed:
gh-104051: fix crash in test_xxtestfuzz with -We (#104052)

kumaraditya303 committed:
GH-103092: port `_asyncio` freelist to module state (#104196)

erlend-aasland committed:
gh-64658: Expand Argument Clinic return converter docs (#104175)

erlend-aasland committed:
gh-68395: Avoid naming conflicts by mangling variable names in Argument Clinic (#104065)

Add all internally used variable names to CLINIC_PREFIXED_ARGS.

erlend-aasland committed:
gh-64631: Test exception messages in cloned Argument Clinic funcs (#104167)

erlend-aasland committed:
gh-64595: Argument Clinic: Touch source file if any output file changed (#104152)

gvanrossum reviewed (commented):
LG, but let's wait until Kumar's feedback has been resolved.

markshannon committed:
GH-96803: Add three C-API functions to make _PyInterpreterFrame less opaque for users of PEP 523. (GH-96849)

jbower-fb committed:
Update asyncio-task.rst

gaogaotiantian committed:
gh-103533: Use PEP 669 APIs for cprofile (GH-103534)

nametkin committed:
gh-69152: Add _proxy_response_headers attribute to HTTPConnection (#26152)

Add _proxy_response_headers attribute to HTTPConnection (#26152)

---------

Co-authored-by: Senthil Kumaran <senthil@python.org>

barneygale committed:
GH-100479: Add `pathlib.PurePath.with_segments()` (GH-103975)

Add `pathlib.PurePath.with_segments()`, which creates a path object from arguments. This method is called whenever a derivative path is created, such as from `pathlib.PurePath.parent`. Subclasses may override this method to share information between path objects.

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

ericsnowcurrently committed:
gh-99113: Share the GIL via PyInterpreterState.ceval.gil (gh-104203)

In preparation for a per-interpreter GIL, we add PyInterpreterState.ceval.gil, set it to the shared GIL for each interpreter, and use that rather than using _PyRuntime.ceval.gil directly.  Note that _PyRuntime.ceval.gil is still the actual GIL.

ericsnowcurrently committed:
gh-104108: Add the Py_mod_multiple_interpreters Module Def Slot (gh-104148)

I'll be adding a value to indicate support for per-interpreter GIL in gh-99114.

ericsnowcurrently committed:
gh-99113: Add Py_MOD_PER_INTERPRETER_GIL_SUPPORTED (gh-104205)

Here we are doing no more than adding the value for Py_mod_multiple_interpreters and using it for stdlib modules.  We will start checking for it in gh-104206 (once PyInterpreterState.ceval.own_gil is added in gh-104204).

erlend-aasland committed:
gh-104146: Remove unused var 'parser_body_declarations' from clinic.py (#104214)

ericsnowcurrently committed:
gh-99113: Add PyInterpreterConfig.own_gil (gh-104204)

We also add PyInterpreterState.ceval.own_gil to record if the interpreter actually has its own GIL.

Note that for now we don't actually respect own_gil; all interpreters still share the one GIL.  However, PyInterpreterState.ceval.own_gil does reflect PyInterpreterConfig.own_gil.  That lie is a temporary one that we will fix when the GIL really becomes per-interpreter.

itamaro committed:
gh-104144: Skip scheduling a done callback if a TaskGroup task completes eagerly (#104140)

Co-authored-by: Carl Meyer <carl@oddbird.net>

jbower-fb committed:
gh-97696: Remove redundant #include (#104216)

Remove "#include cpython/context.h"` from `_asynciomodule.c`.

It's already included in `Python.h`.

vstinner committed:
gh-101819: Prepare to modernize the _io extension (#104178)

* Add references to static types to _PyIO_State:

  * PyBufferedIOBase_Type
  * PyBytesIOBuffer_Type
  * PyIncrementalNewlineDecoder_Type
  * PyRawIOBase_Type
  * PyTextIOBase_Type

* Add the defining class to methods:

  * _io.BytesIO.getbuffer()
  * _io.FileIO.close()

* Add get_io_state_by_cls() function.
* Add state parameter to _textiowrapper_decode()
* _io_TextIOWrapper___init__() now sets self->state before calling
  _textiowrapper_set_decoder().

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

yrro committed:
gh-102215: importlib documentation cleanups

AA-Turner committed:
GH-97950: Use new-style index directive ('builtin') (#104164)

* Uncomment builtin removal in pairindextypes

* Use new-style index directive ('builtin') - C API

* Use new-style index directive ('builtin') - Extending

* Use new-style index directive ('builtin') - Library

* Use new-style index directive ('builtin') - Reference

* Use new-style index directive ('builtin') - Tutorial

sobolevn committed:
gh-104233: Fix "unused variable" warning in `ceval_gil.c` (#104234)

terryjreedy committed:
gh-65772: Clean-up turtle module (#104218)

* Remove the unused, private, and undocumented name `_ver` and
the commented-out `print` call.

* Don't add math functions to `__all__`.  Beginners should learn
to `import math` to access them.

* Gregor Lindel, who wrote this version of turtle, dropped plans
to implement turtle on another toolkit at least a decade ago.
Drop `_dot` code preparing for this, but add a hint comment.

* `_Screen` is meant to be a singleton class.  To enforce that,
it needs either a `__new__` that returns the singleton or
`else...raise` in `__iter__`.  Merely removing the `if` clauses
as suggested might break something if a user were to call `_Screen`
directly.  Leave the code alone until a problem is evident.

* Turtledemo injects into _Screen both _root and _canvas,
configured as it needs them to be.  Making _canvas an `__init__`
option would require skipping some but not all of the lines under
'if _Screen._canvas is None:`.  Leave working code alone.

terryjreedy committed:
Rewrite the turtledemo makeGraphFrame method (#104224)

Replace `self._canvas` and `self.scanvas`, both bound to `canvas`,
with `self.canvas, which is accessed in other methods.
Replace `_s_` with `screen` and `_s_._canvas` with `canvas`.
Add a comment explaining the unorthodox use of
function turtle.Screen and singleton class turtle._Screen.

itamaro committed:
gh-104144: Optimize gather to finish eagerly when all futures complete eagerly (#104138)

kumaraditya303 commented on Doc/library/asyncio-task.rst:
```diff
@@ -1153,6 +1153,11 @@ Task Object
 
       Return the coroutine object wrapped by the :class:`Task`.
 
+      .. note::
```
Add a versionchanged directive.

jbower-fb replied:
Resolved in the new PR https://github.com/python/cpython/pull/104304.

gvanrossum commented on Doc/library/asyncio-task.rst:
```diff
@@ -1153,6 +1153,11 @@ Task Object
+      .. note::
+
+        This will return ``None`` for Tasks which have already
+        completed eagerly.
```
I feel this requires a link to the eager task factory stuff.

kumaraditya303 replied:
Good idea.

jbower-fb replied:
Resolved in the new PR https://github.com/python/cpython/pull/104304

AlexWaygood committed:
gh-90953: Emit deprecation warnings for `ast` features deprecated in Python 3.8 (#104199)

`ast.Num`, `ast.Str`, `ast.Bytes`, `ast.Ellipsis` and `ast.NameConstant` now all emit deprecation warnings on import, access, instantation or `isinstance()` checks.

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

barneygale committed:
GH-103548: Improve performance of `pathlib.Path.[is_]absolute()` (GH-103549)

Improve performance of `pathlib.Path.absolute()` and `cwd()` by joining paths only when necessary. Also improve
performance of `PurePath.is_absolute()` on Posix by skipping path parsing and normalization.

erlend-aasland committed:
gh-101819: Remove unused 'locale_module' from _io state (#104246)

The locale module reference was introduced by 932ff8368 in 2013,
and rendered unused by 710e82630 (gh-23050) in 2020.

ericsnowcurrently committed:
gh-99113: Add a check for Py_MOD_PER_INTERPRETER_GIL_SUPPORTED (gh-104206)

Py_MOD_PER_INTERPRETER_GIL_SUPPORTED is a new supported value for Py_mod_multiple_interpreters, added in gh-104205.

ericsnowcurrently committed:
gh-99113: Make Sure the GIL is Acquired at the Right Places (gh-104208)

This is a pre-requisite for a per-interpreter GIL.  Without it this change isn't strictly necessary.  However, there is no real downside otherwise.

arhadthedev committed:
gh-101640: Make argparse _print_message catch any write error (#101802)

* In particular, don't exit when trying to print to stderr = None.
* Add tests

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

itamaro committed:
gh-104254: Document the optional keyword-only "context" argument to Task constructor (#104251)

(This was added in 3.11. It was already documented for `create_task()`, but not for `Task()`.)

tomasr8 committed:
gh-103886: Improve `builtins.__doc__` (#104179)

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

itamaro committed:
gh-97696: Use `PyObject_CallMethodNoArgs` and inline is_loop_running check in `_asyncio` (#104255)

belm0 committed:
gh-104018: remove unused format "z" handling in string formatfloat() (#104107)

This is a cleanup overlooked in PR #104033.

ymki4360 committed:
Re-enable commented-out test in test_generators.py (#104130)

erlend-aasland committed:
gh-101819: Port _io.PyIncrementalNewlineDecoder_Type to heap type (#104249)

erlend-aasland committed:
gh-101819: Adapt _io.PyWindowsConsoleIO_Type to heap type (#104197)

erlend-aasland committed:
gh-100370: fix OverflowError in sqlite3.Connection.blobopen for 32-bit builds (#103902)

erlend-aasland committed:
gh-101819: Port _io.PyBytesIOBuffer_Type to heap type (#104264)

barneygale committed:
GH-100479: Fix pathlib test failure on WASI (#104215)

AlexWaygood committed:
gh-103193: cache calls to `inspect._shadowed_dict` in `inspect.getattr_static` (#104267)

Co-authored-by: Carl Meyer <carl@oddbird.net>

iritkatriel committed:
gh-104240: make _PyCompile_CodeGen support different compilation modes (#104241)

barneygale committed:
GH-89812: Churn `pathlib.Path` methods (GH-104243)

Re-arrange `pathlib.Path` methods in source code. No other changes.

The methods are arranged as follows:

1. `stat()` and dependants (`exists()`, `is_dir()`, etc)
2. `open()` and dependants (`read_text()`, `write_bytes()`, etc)
3. `iterdir()` and dependants (`glob()`, `walk()`, etc)
4. All other `Path` methods

This patch prepares the ground for a new `_AbstractPath` class, which will
support the methods in groups 1, 2 and 3 above. By churning the methods
here, subsequent patches will be easier to review and less likely to break
things.

art049 committed:
gh-103650: Fix perf maps address format (#103651)

barneygale committed:
GH-102613: Improve performance of `pathlib.Path.rglob()` (GH-104244)

Stop de-duplicating results in `_RecursiveWildcardSelector`. A new
`_DoubleRecursiveWildcardSelector` class is introduced which performs
de-duplication, but this is used _only_ for patterns with multiple
non-adjacent `**` segments, such as `path.glob('**/foo/**')`. By avoiding
the use of a set, `PurePath.__hash__()` is not called, and so paths do not
need to be stringified and case-normalised.

Also merge adjacent '**' segments in patterns.

chgnrdv committed:
gh-104265 Disallow instantiation of `_csv.Reader` and `_csv.Writer` (#104266)

erlend-aasland committed:
gh-64660: Don't hardcode Argument Clinic return converter result variable name (#104200)

buraksaler committed:
gh-104273: Remove redundant len() calls in argparse function (#104274)

msprotz committed:
gh-99108: Replace SHA3 implementation HACL* version (#103597)

Replaces our built-in SHA3 implementation with a verified one from the HACL* project.

This implementation is used when OpenSSL does not provide SHA3 or is not present.

3.11 shiped with a very slow tiny sha3 implementation to get off of the <=3.10 reference implementation that wound up having serious bugs. This brings us back to a reasonably performing built-in implementation consistent with what we've just replaced our other guaranteed available standard hash algorithms with: code from the HACL* project.

---------

Co-authored-by: Gregory P. Smith <greg@krypto.org>

itamaro committed:
gh-97696: Improve and fix documentation for asyncio eager tasks (#104256)

hauntsaninja committed:
gh-102500: Remove mention of bytes shorthand (#104281)

The bytes shorthand was removed in PEP 688:
https://peps.python.org/pep-0688/#no-special-meaning-for-bytes

I also remove the reference to `collections.abc.ByteString`, since that
object is deprecated (#91896) and has different semantics (#102092)

hugovk committed:
Trim trailing whitespace and test on CI (#104275)

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

AlexWaygood committed:
gh-103193: Improve `getattr_static` test coverage (#104286)

AA-Turner committed:
GH-104145: Use fully-qualified cross reference types for the bisect module (#104172)

msprotz committed:
gh-99108: fix typo in Modules/Setup (#104293)

case sensitive filename

JelleZijlstra committed:
gh-104223: Fix issues with inheriting from buffer classes (#104227)

Co-authored-by: Kumar Aditya <59607654+kumaraditya303@users.noreply.github.com>

CCLDArjun committed:
gh-89550: Buffer GzipFile.write to reduce execution time by ~15% (#101251)

Use `io.BufferedWriter` to buffer gzip writes.

---------

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>
Co-authored-by: Gregory P. Smith <greg@krypto.org>

AA-Turner committed:
GH-104284: Fix documentation gettext build (#104296)

jbower-fb committed:
get_coro versionchanged info and cross-reference
